src/kafka-util/src/client.rs

```diff
@@ -323,13 +343,75 @@ impl<C> TunnelingClientContext<C> {
     /// and restarts.
     pub async fn add_broker_rewrite(&self, broker: BrokerAddr, rewrite: BrokerRewrite) {
         let mut rewrites = self.rewrites.lock().await;
-        rewrites.insert(broker, Arc::new(move || rewrite.clone()));
+        rewrites.insert(
+            broker,
+            RewriteStatus::Rewritten(None, Arc::new(move || rewrite.clone())),
+        );
     }
 
     /// Returns a reference to the wrapped context.
     pub fn inner(&self) -> &C {
         &self.inner
     }
+
+    /// Returns a _consolidated_ `SshTunnelStatus` that communicates the status
+    /// of all active ssh tunnels `self` knows about.
+    pub async fn tunnel_statuses(&self) -> SshTunnelStatus {
+        // Pull out all the rewrites, temporarily.
+        let current_rewrites: Vec<_> = self
+            .rewrites
+            .lock()
+            .await
+            .iter()
+            .map(|(broker, status)| (broker.clone(), status.clone()))
+            .collect();
+
+        // First, get the statuses from broken _creation_ of ssh tunnels.
+        let mut ssh_statuses: BTreeMap<_, _> = current_rewrites
+            .iter()
+            .filter_map(|(broker, status)| match status {
+                RewriteStatus::CreationErrored(e) => Some((
+                    SshTunnelKey {
+                        config: self.default_tunnel.clone()?,
+                        remote_host: broker.host.clone(),
+                        remote_port: broker.port.parse().unwrap(),
+                    },
+                    SshTunnelStatus::Errored(e.clone()),
+                )),
+                _ => None,
+            })
+            .collect();
+
+        // Then check the statuses of all the existing tunnels.
+        ssh_statuses.extend(
+            futures::future::join_all(current_rewrites.iter().filter_map(
+                |(_, status)| match status {
+                    RewriteStatus::Rewritten(Some(s), _) => {
+                        let key = s.key.clone();
+                        Some(async { (key, s.check_status().await) })
+                    }
+                    _ => None,
+                },
+            ))
+            .await,
+        );
+
+        // Then consolidate them.
+        let mut status_to_emit = SshTunnelStatus::Running;
+        for (_, status) in ssh_statuses {
+            match (&mut status_to_emit, status) {
+                (SshTunnelStatus::Running, SshTunnelStatus::Errored(e)) => {
+                    status_to_emit = SshTunnelStatus::Errored(e);
+                }
+                (SshTunnelStatus::Errored(ref mut error), SshTunnelStatus::Errored(e)) => {
+                    error.push_str(&e);
```

Comment:
Am I reading this right? We just concat the errors together with no separators? Won't that look quite ugly?

Reply:
oh damn, i just rebased this wrong